cpp/src/parquet/encoding_test.cc

```diff
@@ -862,5 +862,215 @@ TEST_F(DictEncoding, CheckDecodeIndicesNoNulls) {
+    int num_decoded = decoder->Decode(decoded_data.data(), step);
+    ASSERT_EQ(step, num_decoded);
+    for (int j = 0; j < step; ++j) {
+      const uint32_t assembled_value =
+          static_cast<uint32_t>(data[i + j]) |
+          (static_cast<uint32_t>(data[(i + j) + num_values]) << 8U) |
+          (static_cast<uint32_t>(data[(i + j) + num_values * 2]) << 16U) |
+          (static_cast<uint32_t>(data[(i + j) + num_values * 3]) << 24U);
+      const float assembled_value_as_float =
+          *reinterpret_cast<const float*>(&assembled_value);
+      ASSERT_EQ(assembled_value_as_float, decoded_data[j]);
+    }
+  }
+}
+
+// Check that an encode-decode pipeline produces the original small input.
+// This small-input test is added to ease debugging in case of changes to
+// the encoder/decoder implementation.
+TEST(ByteStreamSplitEncodeDecode, SmallInput) {
+  std::unique_ptr<TypedEncoder<FloatType>> encoder =
+      MakeTypedEncoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+
+  const float data[] = {-166.166f, -0.2566f, .0f, 322.0f, 178888.189f};
+  const int num_values = sizeof(data) / sizeof(data[0U]);
+  encoder->Put(data, num_values);
+
+  std::shared_ptr<Buffer> buffer = encoder->FlushValues();
+
+  std::unique_ptr<TypedDecoder<FloatType>> decoder =
+      MakeTypedDecoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+  decoder->SetData(num_values, buffer->mutable_data(), static_cast<int>(buffer->size()));
+
+  std::vector<float> decoded_data(num_values);
+  int num_decoded_elements = decoder->Decode(decoded_data.data(), num_values);
+  ASSERT_EQ(num_values, num_decoded_elements);
+
+  for (size_t i = 0U; i < decoded_data.size(); ++i) {
+    ASSERT_EQ(data[i], decoded_data[i]);
+  }
+}
+
+// Test that the encode-decode pipeline can handle big 32-bit FP input.
+TEST(ByteStreamSplitEncodeDecode, BigInputFloat) {
+  TestEncodeDecodeWithBigInput<FloatType>();
+}
+
+// Test that the encode-decode pipeline can handle big 64-bit FP input.
+TEST(ByteStreamSplitEncodeDecode, BigInputDouble) {
+  TestEncodeDecodeWithBigInput<DoubleType>();
+}
+
+TEST(ByteStreamSplitEncodeDecode, InvalidDataTypes) {
+  // First check encoders.
+  EXPECT_FALSE(TestCreatingByteStreamSplitEncoder<Int32Type>());
+  EXPECT_FALSE(TestCreatingByteStreamSplitEncoder<Int64Type>());
+  EXPECT_FALSE(TestCreatingByteStreamSplitEncoder<Int96Type>());
+  EXPECT_FALSE(TestCreatingByteStreamSplitEncoder<BooleanType>());
+  EXPECT_FALSE(TestCreatingByteStreamSplitEncoder<ByteArrayType>());
+  EXPECT_FALSE(TestCreatingByteStreamSplitEncoder<FLBAType>());
+
+  // Then check decoders.
+  EXPECT_FALSE(TestCreatingByteStreamSplitDecoder<Int32Type>());
+  EXPECT_FALSE(TestCreatingByteStreamSplitDecoder<Int64Type>());
+  EXPECT_FALSE(TestCreatingByteStreamSplitDecoder<Int96Type>());
+  EXPECT_FALSE(TestCreatingByteStreamSplitDecoder<BooleanType>());
+  EXPECT_FALSE(TestCreatingByteStreamSplitDecoder<ByteArrayType>());
+  EXPECT_FALSE(TestCreatingByteStreamSplitDecoder<FLBAType>());
```

Comment:
It would be less verbose to have this helper function return void and use `ASSERT_THROW` instead of implementing that by hand

Reply:
Ok, did not know of ASSERT_THROW. Rewrote the code.